CTAS statement that fails still seems to create a descriptor

**Describe the problem**

The following sql statement taken from `pkg/sql/logictest/testdata/logic_test/create_as` fails but it seems to have still created a table named `foo7`.

```sql
statement error generate insert row: null value in column "x" violates not-null constraint
CREATE TABLE foo7 (x PRIMARY KEY) AS VALUES (1), (NULL);
```

**To Reproduce**

1. Check out this branch: https://github.com/spaskob/cockroach/tree/validate-builtin
2. Run logictest `make testlogic PKG=./pkg/sql TESTS=TestLogic/local-vec-off/create_as`
3. It should fail with error message:
```
        --- FAIL: TestLogic/local-vec-off/create_as (2.17s)
            logic.go:2709:
                testdata/logic_test/create_as:360: error while processing
            logic.go:2709: descriptor validation failed:
                desc &tabledesc.Mutable{
                    Immutable: tabledesc.Immutable{
                        TableDescriptor: descpb.TableDescriptor{
                            Name:                    "foo7",
...
                }
                : parentID 52 does not exist
```

--- a/pkg/sql/schema_changer.go
+++ b/pkg/sql/schema_changer.go
@@ -822,45 +822,64 @@ func (sc *SchemaChanger) rollbackSchemaChange(ctx context.Context, err error) er
 	// Check if the target table needs to be cleaned up at all. If the target
 	// table was in the ADD state and the schema change failed, then we need to
 	// clean up the descriptor.
-	desc, err := sc.getTargetDescriptor(ctx)
-	if err != nil {
-		return err
-	}
-	if tblDesc, ok := desc.(catalog.TableDescriptor); ok && tblDesc.GetState() == descpb.DescriptorState_ADD {
-		// Delete the names in use by this descriptor.
-		if err := sc.db.Txn(ctx, func(ctx context.Context, txn *kv.Txn) error {
-			return catalogkv.RemoveObjectNamespaceEntry(
-				ctx,
-				txn,
-				sc.execCfg.Codec,
-				tblDesc.GetParentID(),
-				tblDesc.GetParentSchemaID(),
-				tblDesc.GetName(),
-				false, /* kvTrace */
-			)
-		}); err != nil {
+	var cleanupJob *jobs.StartableJob
+	if err := sc.txn(ctx, func(ctx context.Context, txn *kv.Txn, descsCol *descs.Collection) error {
+		scTable, err := descsCol.GetMutableTableVersionByID(ctx, sc.descID, txn)
+		if err != nil {
 			return err
 		}
-		// Now start a GC job for the table.
-		if err := startGCJob(
+		if !scTable.Adding() {
+			return nil
+		}
+
+		b := txn.NewBatch()
+		scTable.SetDropped()
+		if err := descsCol.WriteDescToBatch(ctx, false /* kvTrace */, scTable, b); err != nil {
+			return err
+		}
+		catalogkv.WriteObjectNamespaceEntryRemovalToBatch(
 			ctx,
-			sc.db,
-			sc.jobRegistry,
-			sc.job.Payload().Username,
-			"ROLLBACK OF"+sc.job.Payload().Description,
+			b,
+			sc.execCfg.Codec,
+			scTable.GetParentID(),
+			scTable.GetParentSchemaID(),
+			scTable.GetName(),
+			false, /* kvTrace */
+		)
+
+		// Queue a GC job.
+		jobRecord := CreateGCJobRecord(
+			"ROLLBACK OF "+sc.job.Payload().Description,
+			sc.job.Payload().Description,
 			jobspb.SchemaChangeGCDetails{
 				Tables: []jobspb.SchemaChangeGCDetails_DroppedID{
 					{
-						ID:       tblDesc.GetID(),
+						ID:       scTable.GetID(),
 						DropTime: timeutil.Now().UnixNano(),
 					},
 				},
 			},
-		); err != nil {
+		)
+		job, err := sc.jobRegistry.CreateStartableJobWithTxn(ctx, jobRecord, txn, nil /* resultsCh */)
+		if err != nil {
 			return err
 		}
+		cleanupJob = job
+		return txn.Run(ctx, b)
+	}); err != nil {
+		if cleanupJob != nil {
+			if rollbackErr := cleanupJob.CleanupOnRollback(ctx); rollbackErr != nil {
+				log.Warningf(ctx, "failed to clean up job: %v", rollbackErr)
+			}
+		}
+		return err
 	}
-
+	if cleanupJob != nil {
+		if _, err := cleanupJob.Start(ctx); err != nil {
+			log.Warningf(ctx, "starting job %d failed with error: %v", *cleanupJob.ID(), err)
+		}
+		log.VEventf(ctx, 2, "started job %d", *cleanupJob.ID())
+	}
 	return nil
 }
 

--- a/pkg/sql/schema_changer_test.go
+++ b/pkg/sql/schema_changer_test.go
@@ -484,6 +484,20 @@ func TestRollbackOfAddingTable(t *testing.T) {
 	_, err = sqlDB.Exec(`CREATE MATERIALIZED VIEW d.v AS SELECT 1`)
 	require.EqualError(t, err, "pq: boom")
 
+	// Get the view descriptor we just created and verify that it's in the
+	// dropping state. We're unable to access the descriptor via the usual means
+	// because catalog.FilterDescriptorState filters out tables in the ADD state,
+	// and once we move the table to the DROP state we also remove the namespace
+	// entry. So we just get the most recent descriptor.
+	var descBytes []byte
+	row := sqlDB.QueryRow(`SELECT descriptor FROM system.descriptor ORDER BY id DESC LIMIT 1`)
+	require.NoError(t, row.Scan(&descBytes))
+	var desc descpb.Descriptor
+	require.NoError(t, protoutil.Unmarshal(descBytes, &desc))
+	viewDesc := desc.GetTable() //nolint:descriptormarshal
+	require.Equal(t, "v", viewDesc.GetName(), "read a different descriptor than expected")
+	require.Equal(t, descpb.DescriptorState_DROP, viewDesc.GetState())
+
 	// The view should be cleaned up after the failure, so we should be able
 	// to create a new view with the same name.
 	_, err = sqlDB.Exec(`CREATE MATERIALIZED VIEW d.v AS SELECT 1`)

